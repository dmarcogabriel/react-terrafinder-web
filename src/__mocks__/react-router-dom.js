const originalModule = jest.requireActual('react-router-dom');

module.exports = {
  ...originalModule,
  useHistory: () => ({
    history: { push: jest.fn() },
  }),
};
