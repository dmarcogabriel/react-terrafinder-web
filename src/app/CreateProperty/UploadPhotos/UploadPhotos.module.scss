@import 'styles';

.uploadPhotos {
  padding: 0 2rem;

  @media ($desktop) {
    padding: 0;
  }

  color: $gray-dark;

  > h2 {
    text-align: center;
    font-weight: normal;
    margin: 1rem 0;
    font-size: 1.5rem;
  }

  .dropZone {
    border: 2px dashed $gray-dark;
    border-radius: 0.2rem;
    padding: 3rem 0;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    color: $gray-dark;

    > input {
      visibility: hidden;
    }

    .button {
      background: $gray-light;
      color: $gray-dark;
      margin-bottom: 2rem;
    }

    > span {
      font-size: 0.8rem;
      font-weight: bold;
    }

    .icon {
      margin: 2rem 0;
    }

    .previewImages {
      @media ($desktop) {
        display: grid;
        grid-template-columns: repeat(3, 1fr);
        gap: 1rem;
      }

      padding: 0 1rem;

      > div {
        position: relative;
        margin: 1rem 0;

        @media ($desktop) {
          margin: 0;
        }

        > img {
          width: 100%;
          max-height: 15rem;

          @media ($desktop) {
            max-height: none;
          }
        }

        .removeImageIcon {
          position: absolute;
          top: 0.5rem;
          right: 0.5rem;
          color: $orange;
          background: black;
          border-radius: 100%;
          cursor: pointer;
        }
      }
    }
  }
}
