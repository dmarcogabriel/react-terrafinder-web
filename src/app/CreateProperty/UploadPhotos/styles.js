import styled, { css } from 'styled-components';
import Button from 'common/components/atm/Button';

export const UploadButton = styled(Button)`
  margin: 2rem 0;
  ${({ theme: { color } }) => css`
    background: ${color.gray.lt};
    color: ${color.gray.dk};
  `}
`;
