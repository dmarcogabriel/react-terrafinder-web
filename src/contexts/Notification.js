import React, { createContext, useState, useEffect } from 'react';

export const Context = createContext();

export const NotificationProvider = ({ children }) => {
  const [show, setShow] = useState(false);
  const [notification, setNotification] = useState({
    message: '',
    type: 'info',
  });

  const showNotification = (message, type = 'info') => {
    setNotification({ message, type });
    setShow(true);
  };

  useEffect(() => {
    if (show) {
      setTimeout(() => {
        setShow(false);
      }, 5000);
    }
  }, [show]);

  return (
    <Context.Provider
      value={{
        show,
        notification,
        showNotification,
      }}
    >
      {children}
    </Context.Provider>
  );
};
