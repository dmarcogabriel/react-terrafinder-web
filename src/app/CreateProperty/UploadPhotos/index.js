import React, { useState, useRef } from 'react';
import { MdCloudUpload } from 'react-icons/md';
import Button from 'common/components/Button';
import { validateFile } from 'utils/validators';
import { loadBase64Image } from 'utils/fileHelper';
import { AiFillCloseCircle } from 'react-icons/ai';
import { useHistory, useRouteMatch } from 'react-router-dom';
import api from 'services/api';
import { NOTIFICATION_TYPES, useNotification } from 'hooks/useNotification';
import { useUser } from 'hooks/useUser';
import classes from './UploadPhotos.module.scss';
import Navigator from '../Navigator';

export default function UploadPhotos() {
  const [opacity, setOpacity] = useState(1);
  const [selectedFiles, setSelectedFiles] = useState([]);
  const [, setUploading] = useState(false); // todo: show modal
  const history = useHistory();
  const { params } = useRouteMatch();
  const { showNotification } = useNotification();
  const inputRef = useRef();
  const { currentUser } = useUser();

  const handleDrop = async (e) => {
    e.preventDefault();
    setOpacity(1);

    if (selectedFiles.length < 3) {
      const { files } = e.dataTransfer;

      if (files.length) {
        for (let i = 0; i < files.length; i++) {
          if (validateFile(files[i])) {
            const file = files[i];

            file.data = await loadBase64Image(files[i]);

            setSelectedFiles([...selectedFiles, file]);
          }
        }
      }
    }
  };

  const handleDragOver = (e) => {
    e.preventDefault();
    setOpacity(0.7);
  };

  const handleDragLeave = (e) => {
    e.preventDefault();
    setOpacity(1);
  };

  const removeImage = (index) => {
    setSelectedFiles(selectedFiles.filter((_, i) => i !== index));
  };

  const goBack = () => history.replace('/dashboard');

  const uploadPhotos = async () => {
    try {
      setUploading(true);

      const data = new FormData();

      selectedFiles.forEach((file, i) => {
        data.append(`photo${i + 1}`, file);
      });

      await api.post(`property/upload-photos/${params.id}`, data, {
        headers: {
          'Content-Type': 'multipart/form-data',
          'x-access-token': currentUser.token,
        },
      });

      setUploading(false);
      history.replace('/dashboard');
      showNotification(
        'Fotos enviadas com sucesso!',
        NOTIFICATION_TYPES.SUCCESS
      );
    } catch (error) {
      setUploading(false);
      showNotification(
        'Ocorreu um erro ao enviar fotos!',
        NOTIFICATION_TYPES.ERROR
      );
    }
  };

  const openGalery = () => inputRef.current.click();

  const handleChangeFileInput = async (e) => {
    if (selectedFiles.length < 3) {
      const [file] = e.target.files;

      file.data = await loadBase64Image(file);

      setSelectedFiles((old) => [...old, file]);
    }
  };

  return (
    <div className={classes.uploadPhotos}>
      <h2>Imagens do Imóvel</h2>

      <div
        className={classes.dropZone}
        onDrop={handleDrop}
        onDragOver={handleDragOver}
        onDragLeave={handleDragLeave}
        style={{ opacity }}
      >
        <p className={classes.helperText}>
          Faça o upload ou araste a imagem aqui
        </p>

        <input
          onChange={handleChangeFileInput}
          ref={inputRef}
          accept="image/png, image/jpg"
          type="file"
        />

        <Button className={classes.button} onClick={openGalery}>
          Upload de Imagens
        </Button>

        <span>JPG, JPEG, PNG</span>

        {!selectedFiles.length && (
          <div className={classes.icon}>
            <MdCloudUpload size={100} />
          </div>
        )}

        <div className={classes.previewImages}>
          {selectedFiles.map((file, i) => (
            <div key={String(i)}>
              <button type="button" onClick={() => removeImage(i)}>
                <AiFillCloseCircle
                  size={26}
                  className={classes.removeImageIcon}
                />
              </button>

              <img src={file.data} alt={file.name} />
            </div>
          ))}
        </div>
      </div>

      <Navigator
        nextButtonText="Fazer Upload"
        onBack={goBack}
        onNext={uploadPhotos}
      />
    </div>
  );
}
