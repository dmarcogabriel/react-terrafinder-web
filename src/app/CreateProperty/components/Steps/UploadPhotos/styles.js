import { styled } from '@mui/material/styles';
import { Button, Typography } from '@mui/material';
import { grey } from '@mui/material/colors';

export const LabelText = styled(Typography)({
  color: grey[500],
});

export const UploadButton = styled(Button)({
  borderRadius: '20px',
});
