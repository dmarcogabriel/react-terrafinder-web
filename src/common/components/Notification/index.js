import React from 'react';
import { useNotification } from 'hooks/useNotification';
import './styles.scss';
import cn from 'classnames';

export default function Notification() {
  const { notification, show } = useNotification();

  const style = {
    opacity: show ? 0.8 : 0,
    zIndex: show ? -50 : 5000,
  };

  return (
    <div
      data-testid="notification"
      className={cn('Notification__container', notification.type)}
      style={style}
    >
      <p data-testid="notificationMessage">{notification.message}</p>
    </div>
  );
}
