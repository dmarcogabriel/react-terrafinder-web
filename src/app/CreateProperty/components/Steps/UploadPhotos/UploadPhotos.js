import React, { useState, useRef } from 'react';
import { MdCloudUpload } from 'react-icons/md';
import { validateFile } from 'utils/validators';
import { loadBase64Image } from 'utils/fileHelper';
import { AiFillCloseCircle } from 'react-icons/ai';
import { useHistory, useRouteMatch, useLocation } from 'react-router-dom';
import { NOTIFICATION_TYPES, useNotification } from 'hooks/useNotification';
import { Modal, useModal } from 'common/components';
import { Typography, Button, Box } from '@mui/material';
import { useUser } from 'hooks/useUser';
import { CameraAltOutlined as CameraIcon } from '@mui/icons-material';
import classes from './UploadPhotos.module.scss';
import Navigator from '../../Navigator';
import { UploadButton, LabelText } from './styles';
import { useNewProperty } from '../../../NewPropertyContext';
import StepTitle from '../../StepTtitle';

export const UploadPhotos = () => {
  const [opacity, setOpacity] = useState(1);
  const [selectedFiles, setSelectedFiles] = useState([]);
  const [, setUploading] = useState(false); // todo: show modal
  const history = useHistory();
  const { params } = useRouteMatch();
  const { showNotification } = useNotification();
  const inputRef = useRef();
  const { state } = useLocation();
  const { open, triggerModal } = useModal();
  const { currentUser } = useUser();
  const { nextStep, setPhotos } = useNewProperty();

  const handleDrop = async (e) => {
    e.preventDefault();
    setOpacity(1);

    if (selectedFiles.length < 3) {
      const { files } = e.dataTransfer;

      if (files.length) {
        for (let i = 0; i < files.length; i++) {
          if (validateFile(files[i])) {
            const file = files[i];

            file.data = await loadBase64Image(files[i]);

            setSelectedFiles([...selectedFiles, file]);
          }
        }
      }
    }
  };

  const handleDragOver = (e) => {
    e.preventDefault();
    setOpacity(0.7);
  };

  const handleDragLeave = (e) => {
    e.preventDefault();
    setOpacity(1);
  };

  const removeImage = (index) => {
    setSelectedFiles(selectedFiles.filter((_, i) => i !== index));
  };

  // const uploadPhotos = async () => {
  //   try {
  //     setUploading(true);

  //     const data = new FormData();

  //     selectedFiles.forEach((file, i) => {
  //       data.append(`photo${i + 1}`, file);
  //     });

  //     await api.put(
  //       `properties/upload-photos/${params.id}`,
  //       { ...data, userId: currentUser._id },
  //       {
  //         headers: {
  //           'Content-Type': 'multipart/form-data',
  //         },
  //       }
  //     );

  //     setUploading(false);
  //     history.replace('/create-property/review', {
  //       plan: state.plan,
  //       propertyId: params.id,
  //     });
  //     showNotification(
  //       'Fotos enviadas com sucesso!',
  //       NOTIFICATION_TYPES.SUCCESS
  //     );
  //   } catch (error) {
  //     setUploading(false);
  //     showNotification(
  //       'Ocorreu um erro ao enviar fotos!',
  //       NOTIFICATION_TYPES.ERROR
  //     );
  //   }
  // };

  const openGalery = () => inputRef.current.click();

  const handleChangeFileInput = async (e) => {
    if (selectedFiles.length < 3) {
      const [file] = e.target.files;

      file.data = await loadBase64Image(file);

      setSelectedFiles((old) => [...old, file]);
    }
  };

  const handleSendPhotosLater = () =>
    history.replace('/create-property/review');

  const handleNext = () => {
    if (selectedFiles.length) {
      setPhotos(selectedFiles);
      history.replace('/create-property/review');
      nextStep();
    } else triggerModal();
  };

  return (
    <>
      <div className={classes.uploadPhotos}>
        <Box sx={{ textAlign: 'center' }}>
          <CameraIcon fontSize="large" color="success" />
          <StepTitle>Conteúdos do Imóvel</StepTitle>
        </Box>

        <div
          data-testid="dropZone"
          className={classes.dropZone}
          onDrop={handleDrop}
          onDragOver={handleDragOver}
          onDragLeave={handleDragLeave}
          style={{ opacity }}
        >
          <LabelText sx={{ my: 1 }}>
            Faça o upload ou arraste a imagem aqui
          </LabelText>

          <input
            data-testid="uploadInput"
            onChange={handleChangeFileInput}
            ref={inputRef}
            accept="image/*"
            type="file"
          />

          <UploadButton
            data-testid="uploadButton"
            color="grey"
            variant="contained"
            sx={{ my: 1, px: 5 }}
            onClick={openGalery}
          >
            Upload de Imagens
          </UploadButton>

          <Typography variant="body2" sx={{ fontWeight: 'bold', my: 1 }}>
            JPG, JPEG, PNG
          </Typography>

          {!selectedFiles.length && (
            <div className={classes.icon}>
              <MdCloudUpload size={100} />
            </div>
          )}

          <div className={classes.previewImages}>
            {selectedFiles.map((file, i) => (
              <div data-testid={`image-${i}`} key={String(i)}>
                <button
                  data-testid={`delete-${i}`}
                  type="button"
                  onClick={() => removeImage(i)}
                >
                  <AiFillCloseCircle
                    size={26}
                    className={classes.removeImageIcon}
                  />
                </button>

                <img src={file.data} alt={file.name} />
              </div>
            ))}
          </div>
        </div>

        <Navigator
          nextButtonText={
            selectedFiles.length ? 'Próximo' : 'Enviar mais tarde'
          }
          onBack={history.goBack}
          onNext={handleNext}
        />
      </div>
      <Modal show={open} onClose={triggerModal}>
        <Typography sx={{ mb: 2 }}>
          Tem certeza de que deseja enviar fotos mais tarde?
        </Typography>

        <Box
          sx={{
            display: 'flex',
            flexDirection: { xs: 'column', md: 'row' },
            justifyContent: 'space-between',
          }}
        >
          <Button
            color="info"
            variant="contained"
            sx={{ my: 1 }}
            onClick={triggerModal}
          >
            Voltar para enviar
          </Button>
          <Button sx={{ my: 1 }} color="error" onClick={handleSendPhotosLater}>
            Enviar mais tarde
          </Button>
        </Box>
      </Modal>
    </>
  );
};
