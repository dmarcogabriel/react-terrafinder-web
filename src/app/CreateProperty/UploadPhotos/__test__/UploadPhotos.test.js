import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { NotificationProvider } from 'contexts/Notification';
import { UserProvider } from 'contexts/User';
import UploadPhotos from '..';
import mockImage from './mockImage.jpg';

const Comp = (props) => (
  <NotificationProvider>
    <UserProvider>
      <Router>
        <UploadPhotos {...props} />
      </Router>
    </UserProvider>
  </NotificationProvider>
);

describe('<UploadPhotos />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Comp />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should open galery', () => {
    const { getByTestId } = render(<Comp />);

    const uploadInput = getByTestId('uploadInput');
    const mockFn = jest.spyOn(uploadInput, 'click');

    const uploadButton = getByTestId('uploadButton');
    fireEvent.click(uploadButton);

    expect(mockFn).toHaveBeenCalled();
  });

  // todo: test on change
  // it('should fire change file input', () => {
  //   const { getByTestId } = render(<Comp />);

  //   const uploadInput = getByTestId('uploadInput');
  //   fireEvent.change(uploadInput, { target: { files: [mockImage] } });

  //   const image = getByTestId('image-0');

  //   expect(image).toBeInTheDocument();
  // });

  describe('drop zone', () => {
    it('should change drop zone opacity', () => {
      const { getByTestId } = render(<Comp />);

      const dropZone = getByTestId('dropZone');
      fireEvent.dragOver(dropZone);

      expect(dropZone.style.opacity).toEqual('0.7');
    });

    it('should change drop zone opacity to default', () => {
      const { getByTestId } = render(<Comp />);

      const dropZone = getByTestId('dropZone');
      fireEvent.dragOver(dropZone);
      fireEvent.dragLeave(dropZone);

      expect(dropZone.style.opacity).toEqual('1');
    });
  });

  it('matches snapshot', () => {
    const tree = renderer.create(<Comp />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
