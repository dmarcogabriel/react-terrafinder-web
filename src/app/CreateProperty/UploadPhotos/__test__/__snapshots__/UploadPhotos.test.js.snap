// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UploadPhotos /> matches snapshot 1`] = `
<div>
  <h2>
    Imagens do Imóvel
  </h2>
  <div
    data-testid="dropZone"
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <p>
      Faça o upload ou araste a imagem aqui
    </p>
    <input
      accept="image/png, image/jpg"
      data-testid="uploadInput"
      onChange={[Function]}
      type="file"
    />
    <button
      className="button"
      data-testid="uploadButton"
      onClick={[Function]}
      type="button"
    >
      Upload de Imagens
    </button>
    <span>
      JPG, JPEG, PNG
    </span>
    <div>
      <svg
        fill="currentColor"
        height={100}
        size={100}
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width={100}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
        />
      </svg>
    </div>
    <div />
  </div>
  <div>
    <button
      data-testid="back"
      onClick={[Function]}
      type="button"
    >
      Voltar
    </button>
    <button
      className="button"
      data-testid="next"
      onClick={[Function]}
      type="button"
    >
      Fazer Upload
    </button>
  </div>
</div>
`;
